Bluetooth: Mesh: Use relay bufs/pool for brg_cfg

HaavardRei opened the pull request:
Adds support for using relay buffers and advertising set for the subnet bridge feature, even if the relay feature is disabled. Introduces new Kconfig options that allows the user to configure the relay buffer count and advertising sets in the case when the relay feature is disabled.

HaavardRei commented on subsys/bluetooth/mesh/adv.c:
```diff
@@ -52,13 +52,20 @@ K_MEM_SLAB_DEFINE_STATIC(local_adv_pool, sizeof(struct bt_mesh_adv),
 			 CONFIG_BT_MESH_FRIEND_LPN_COUNT, __alignof__(struct bt_mesh_adv));
 #endif
 
+#if (!defined(CONFIG_BT_MESH_RELAY_ADV_SETS) && defined(CONFIG_BT_MESH_BRG_ADV_SETS))
+#define CONFIG_BT_MESH_RELAY_ADV_SETS CONFIG_BT_MESH_BRG_ADV_SETS
+#endif
```
Not sure if this is the best way to go about it, but added this to simplify the expression following line 212

omkar3141 commented on doc/connectivity/bluetooth/api/mesh/brg_cfg.rst:
```diff
@@ -113,13 +113,16 @@ protection to ensure network security. Key considerations to take into account a
+:kconfig:option:`CONFIG_BT_MESH_BRG_ADV_SETS` and :kconfig:option:`CONFIG_BT_MESH_BRG_BUF_COUNT`
+will be used.
+
+To ensure that messages can be retransmitted to all subnetworks, the buffer count option should be
+increased accordingly.
```
The value of the RELAY_BUF_COUNT/BRG_BUF_COUNT will impact the relay/bridge latency directly. The default value is choosen to optimize good latency with lesser packet loss. Therefore, I will strongly suggest to remove this recommendation to avoid adding confusion.

omkar3141 commented on doc/connectivity/bluetooth/api/mesh/brg_cfg.rst:
```diff
@@ -113,13 +113,16 @@ protection to ensure network security. Key considerations to take into account a
-:kconfig:option:`CONFIG_BT_MESH_ADV_BUF_COUNT` option to allow for sufficient buffer space.
+When a message is relayed between subnets by a Subnet Bridge, it is allocated from the relay buffer
+pool and transmitted using the relay advertising sets. These will be used even if the relay feature
+:kconfig:option:`CONFIG_BT_MESH_RELAY` is disabled. If enabled, the relay buffers and advertising
+sets will use the parameters set by :kconfig:option:`CONFIG_BT_MESH_RELAY_ADV_SETS` and
+:kconfig:option:`CONFIG_BT_MESH_RELAY_BUF_COUNT`. If relaying is disabled, the parameters set by
+:kconfig:option:`CONFIG_BT_MESH_BRG_ADV_SETS` and :kconfig:option:`CONFIG_BT_MESH_BRG_BUF_COUNT`
+will be used.
```
May I suggest something else? In my mind, I always assumed that "If relay feature is disabled, SBR will still automatically use CONFIG_BT_MESH_RELAY_ADV_SETS and CONFIG_BT_MESH_RELAY_BUF_COUNT. 

Additional CONFIG_BT_MESH_BRG_ADV_SETS, and CONFIG_BT_MESH_BRG_BUF_COUNT does not really provide a real value. As a product user, you would like to have a similar performance for relay and a bridge, which means ... both of these functions should ideally use a single KConfig options for controlling sets and buffer counts. We can discuss this offline if needed.

HaavardRei replied:
Hmm I guess we could extract `BT_MESH_RELAY_ADV_SETS` and `BT_MESH_RELAY_BUF_COUNT` and let them be visible if `BT_MESH_RELAY` or `BT_MESH_BRG_CFG_SRV` is enabled. I'm not sure where we would place them though, as this might affect the menuconfig. Any suggestions?

HaavardRei replied:
Pushed a suggestion for this.

alxelax reviewed (commented):
Is it really required to have more config options?

What if you add just dependency to the relay buffer\advertiser instances on bridge feature?

Advantages:
It will not require any C code changes. You will not need more KConfig options. There is only one place to configure number of buffers.

Disadvantages:
It requires better documentation with description that you require more buffers if two features are enabled.

My opinion is less entities in both C and KConfig is better than more. It is better maintainable.

alxelax commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -170,6 +170,19 @@ config BT_MESH_RELAY_ADV_SETS
 	  Maximum of simultaneous relay message support. Requires controller support
 	  multiple advertising sets.
 
+config BT_MESH_BRG_ADV_SETS
```
It is quite cumbersome. I set this parameter to correct value and forget about this. Everything works.
After a couple of months, I enable relay feature, and my subnetwork bridge starts dropping messages due to usage of another set of Kconfig options.

HaavardRei replied:
`BT_MESH_RELAY_BUF_COUNT` is under "if `BT_MESH_RELAY`, and `BT_MESH_RELAY_ADV_SETS` depends on it, so as I mentioned to @omkar3141 above we would need to extract and place them somewhere commonly accessible for both the relay and brg module

LingaoM commented:
I previously proposed that the relay advertising set be renamed to a common name, for example: parallel advertising set, which not depends any feature, and can be used by relay, provision and brg_cfg.

HaavardRei commented:
> I previously proposed that the relay advertising set be renamed to a common name, for example: parallel advertising set, which not depends any feature, and can be used by relay, provision and brg_cfg.

IMO it's a bit more invasive than the current changes, but not a bad idea. This would require some restructuring and doc changes to functions/Kconfigs such as `BT_MESH_PB_ADV_USE_RELAY_SETS`, `bt_mesh_adv_relay_ready`, etc. 
I would rather suggest creating a separate issue for this.

HaavardRei commented:
This might pose a problem:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/net.c#L709
If relaying is disabled, the number of retransmits returned will be 0 for bridged messages (https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/cfg.c#L373)
I can't just add a simple `|| CONFIG_BT_MESH_BRG_CFG_SRV` here as `bt_mesh.relay_xmit` is not compiled in. Any suggestions on how to deal with this? (I. e. compiling in retransmits anyway, or adding a separate option, or something else)

alxelax commented:
> This might pose a problem: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/net.c#L709 If relaying is disabled, the number of retransmits returned will be 0 for bridged messages (https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/cfg.c#L373) I can't just add a simple `|| CONFIG_BT_MESH_BRG_CFG_SRV` here as `bt_mesh.relay_xmit` is not compiled in. Any suggestions on how to deal with this? (I. e. compiling in retransmits anyway, or adding a separate option, or something else)

I guess in case of bridging between subnetworks it should be networked xmit.

KyraLengfeld commented on doc/connectivity/bluetooth/api/mesh/brg_cfg.rst:
```diff
@@ -113,13 +113,11 @@ protection to ensure network security. Key considerations to take into account a
-allocated from the advertising buffer instead. In this case, increase the
-:kconfig:option:`CONFIG_BT_MESH_ADV_BUF_COUNT` option to allow for sufficient buffer space.
+When a message is relayed between subnets by a Subnet Bridge, it is allocated from the relay buffer
+pool and transmitted using the relay advertising sets. These will be used even if the relay feature
```
Reading this isolated, I am missing half a sentence what these buffer and advertising sets are used for when MESH_RELAY is disabled... not for relaying, right?

HaavardRei replied:
The relay and bridge functionality is quite similar IMO, with relaying you're a middleman for nodes in the same subnet, but with subnet bridge you're a middleman between subnets. So it's _kind of_ relaying. The `bt_mesh_net_relay` func is used for both: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/net.c#L685

PavelVPV commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -447,6 +451,26 @@ config BT_MESH_MSG_CACHE_SIZE
 	  Setting this value to a very large number can impact the processing time
 	  for each received network PDU and increases RAM footprint proportionately.
 
+config BT_MESH_RELAY_BUF_COUNT
```
Maybe instead move `endif # BT_MESH_RELAY` above `config BT_MESH_RELAY_BUF_COUNT`? Then only changed text will be visible.

HaavardRei committed:
Bluetooth: Mesh: Use net xmit params for bridge

Makes sure that the Network Transmit state is used when using a Subnet
Bridge.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

HaavardRei committed:
Bluetooth: Mesh: Use relay bufs/pool for brg_cfg

Adds support for using relay buffers and advertising set for the subnet
bridge feature, even if the relay feature is disabled.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

HaavardRei commented on doc/connectivity/bluetooth/api/mesh/brg_cfg.rst:
```diff
@@ -113,13 +113,11 @@ protection to ensure network security. Key considerations to take into account a
+pool and transmitted using the relay advertising sets. These will be used even if the relay feature
+:kconfig:option:`CONFIG_BT_MESH_RELAY` is disabled. The number of relay buffers and advertising sets
+are configurable using the :kconfig:option:`CONFIG_BT_MESH_RELAY_BUF_COUNT` and
+:kconfig:option:`CONFIG_BT_MESH_RELAY_ADV_SETS` Kconfig options.
```
Good point, done!

HaavardRei commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -165,11 +165,15 @@ config BT_MESH_RELAY_ADV_SETS
 	  multiple advertising sets.
 
+	  Note that: The Subnet Bridge feature uses the relay advertising sets. If both the Relay
+	  feature and the Subnet Bridge feature is enabled, the added load should be taken into
```
Removed from adv sets!

omkar3141 reviewed (approved):
Now it looks aligned. I have tiny thing about name of the KConfig itself (like calling them as "RELAY_BUF_COUNT" as they are more generic now), but let's leave it for now.
